### PR TITLE
Add readonly mode to APY daemon

### PR DESCRIPTION
**Pull request by ruuda**
I wanted to test some changes in the server part, but have it run against an existing database that a different instance of the daemon is already writing to. So add a read-only mode (`--fetch disable`) that runs only the http server part and not the fetch-prices-from-the-chain part.

I would have liked to have `--fetch` and `--no-fetch` flags instead, but I don’t know how to do this with Clap `#[derive]` macros. I found some discussions about it with workarounds that use the method-chaining builders. So for now let’s make it an option instead.

### CODE REVIEW

**Comment by enriquefynn** [comment]
We could use a boolean flag `--read-only`.
```rust
#[clap(long)]
read_only: bool,
```
What do you think?

**Comment by ruuda** [comment]
Yeah, that’s nicer, I’ll do that.

**Comment by ruuda** [comment]
I switched the flag. I also noticed that the `/apy` endpoint returned a misleading error, I refactored the route parsing to return more helpful errors, and I think it’s a bit simpler like this as well. Please take another look.

File: cli/listener/src/main.rs [hunk 1]
```diff
@@ -529,41 +536,37 @@ fn get_interval_price_request(
...
+        .and_then(|segments| segments.last());
+
+    match final_segment {
+        Some("apy") => {
+            get_date_params(parsed_url.query_pairs()).map(Endpoint::IntervalPriceRequest)
         }
-        "metrics" => Some(ParseUrlResponse::Metrics),
-        _ => None,
+        Some("metrics") => Ok(Endpoint::Metrics),
+        _ => Err(ResponseError::NotFound("Unknown route.")),
```
Comment (enriquefynn) [inline cli/listener/src/main.rs:569]: 👍

**Review by enriquefynn — APPROVED** [review]
LGTM!

File: cli/listener/src/main.rs [hunk 2]
```diff
@@ -529,41 +536,37 @@ fn get_interval_price_request(
...
-        "apy" => {
-            let query_params =
-                form_urlencoded::parse(parsed_url.query()?.as_bytes()).collect::<Vec<_>>();
-            let dates = match get_date_params(query_params) {
-                Ok(res) => res,
-                Err(err) => return Some(ParseUrlResponse::Error(err)),
-            };
-            Some(ParseUrlResponse::IntervalPriceRequest(dates))
+    let base = Url::parse("http://unused.invalid/").expect("Hard-coded value is valid.");
+    let parse_result = Url::options().base_url(Some(&base)).parse(request_url);
```
[the remark this answers is not in the document]
Comment (ruuda) [reply cli/listener/src/main.rs:557]: Not sure if it makes a difference in practice, but it looked like a more "proper" way to do it from the docs.

**Pull request merged**